File: sdk/typescript/animus-plugin-sdk/src/wire.ts
// Newline-delimited JSON-RPC 2.0 transport over a Node Readable/Writable pair
// (stdin/stdout by default).
//
// Mirrors the Rust runtime in `crates/animus-plugin-runtime/src/lib.rs`:
//   - one JSON-RPC frame per line of stdin
//   - replies + server-streaming notifications are newline-terminated JSON
//     written to stdout
//   - parse errors on a single line are logged-and-skipped (do not kill the
//     read loop)
//   - empty/whitespace-only lines are ignored
//
// Stdout is reserved for protocol frames; SDK diagnostics MUST go to stderr.

import type { Readable, Writable } from 'node:stream';
import { stdin as nodeStdin, stdout as nodeStdout } from 'node:process';
import { StringDecoder } from 'node:string_decoder';

import type { RpcId, RpcNotification, RpcRequest, RpcResponse } from './types/index.js';

/** A handler invoked for every parsed inbound JSON-RPC frame.
 *  Return `undefined` for notifications (no response expected). */
export type FrameHandler = (frame: RpcRequest) => Promise<RpcResponse | undefined> | RpcResponse | undefined;

export interface WireOptions {
  /** Readable stream to consume newline-delimited JSON-RPC frames from. Defaults to process.stdin. */
  input?: Readable;
  /** Writable stream to write newline-terminated JSON-RPC frames to. Defaults to process.stdout. */
  output?: Writable;
  /** Logger for invalid frames / errors. Defaults to `console.error`. */
  logger?: (msg: string, err?: unknown) => void;
}

export interface Wire {
  /** Send a JSON-RPC response. */
  sendResponse(response: RpcResponse): Promise<void>;
  /** Send a JSON-RPC server-streaming notification (no `id`). */
  sendNotification(notification: RpcNotification): Promise<void>;
  /** Convenience: build + send a notification from method/params. */
  notify(method: string, params?: unknown): Promise<void>;
  /** Begin consuming the input stream. Resolves when the stream closes (EOF). */
  run(handler: FrameHandler): Promise<void>;
}

const defaultLogger = (msg: string, err?: unknown): void => {
  if (err !== undefined) {
    process.stderr.write(`[animus-plugin-sdk] ${msg}: ${String(err)}\n`);
  } else {
    process.stderr.write(`[animus-plugin-sdk] ${msg}\n`);
  }
};

/**
 * Encode a single JSON-RPC frame as a UTF-8 newline-terminated string.
 * Exposed for tests; production code should go through `Wire.sendResponse` /
 * `Wire.sendNotification`.
 */
export function encodeFrame(frame: RpcResponse | RpcNotification | RpcRequest): string {
  return `${JSON.stringify(frame)}\n`;
}

/**
 * Parse one trimmed line into an `RpcRequest`. Throws on invalid JSON or
 * structurally invalid JSON-RPC (missing `jsonrpc: "2.0"` or `method`).
 */
export function parseFrame(line: string): RpcRequest {
  const value = JSON.parse(line) as unknown;
  if (typeof value !== 'object' || value === null) {
    throw new Error('frame is not a JSON object');
  }
  const obj = value as Record<string, unknown>;
  if (obj.jsonrpc !== '2.0') {
    throw new Error(`unsupported jsonrpc version: ${String(obj.jsonrpc)}`);
  }
  if (typeof obj.method !== 'string' || obj.method.length === 0) {
    throw new Error('frame missing string `method`');
  }
  const frame: RpcRequest = {
    jsonrpc: '2.0',
    method: obj.method,
  };
  if ('id' in obj) {
    frame.id = obj.id as RpcId;
  }
  if ('params' in obj) {
    frame.params = obj.params;
  }
  return frame;
}

class WriteQueue {
  private chain: Promise<void> = Promise.resolve();
  constructor(private readonly output: Writable) {}

  enqueue(payload: string): Promise<void> {
    const next = this.chain.then(
      () =>
        new Promise<void>((resolve, reject) => {
          const ok = this.output.write(payload, (err) => {
            if (err) reject(err);
            else resolve();
          });
          if (!ok) {
            this.output.once('drain', () => {
              // write callback still fires; nothing to do here.
            });
          }
        }),
    );
    // Swallow rejections in the chain so a single write failure doesn't poison
    // the queue; the original caller still sees the rejection.
    this.chain = next.catch(() => undefined);
    return next;
  }
}

/**
 * Create a `Wire` bound to the given (or default stdin/stdout) streams.
 *
 * Reads are line-buffered: a frame ends at the next `\n`. Frames that fail to
 * parse are logged via `options.logger` and skipped, matching the Rust
 * runtime's "tracing::warn + continue" behavior.
 */
export function createWire(options: WireOptions = {}): Wire {
  const input = options.input ?? nodeStdin;
  const output = options.output ?? nodeStdout;
  const log = options.logger ?? defaultLogger;
  const queue = new WriteQueue(output);

  const sendResponse: Wire['sendResponse'] = (response) => queue.enqueue(encodeFrame(response));
  const sendNotification: Wire['sendNotification'] = (notification) =>
    queue.enqueue(encodeFrame(notification));
  const notify: Wire['notify'] = (method, params) => {
    const frame: RpcNotification = params === undefined
      ? { jsonrpc: '2.0', method }
      : { jsonrpc: '2.0', method, params };
    return sendNotification(frame);
  };

  const run: Wire['run'] = (handler) =>
    new Promise<void>((resolve, reject) => {
      let buffer = '';
      let closed = false;
      // `StringDecoder` buffers partial multi-byte UTF-8 sequences across
      // chunk boundaries — `chunk.toString('utf8')` would replace a split
      // codepoint with U+FFFD and corrupt valid JSON containing non-ASCII
      // subject titles.
      const decoder = new StringDecoder('utf8');
      // Serialize request handling. Plugins can still kick off background work
      // inside their handler (and emit streaming notifications via
      // `notify`/`sendNotification`); the dispatch order matches inbound order
      // so request ids are responded to in arrival sequence by default.
      let dispatchChain: Promise<void> = Promise.resolve();

      const handleOne = async (line: string): Promise<void> => {
        const trimmed = line.trim();
        if (trimmed.length === 0) return;
        let frame: RpcRequest;
        try {
          frame = parseFrame(trimmed);
        } catch (err) {
          log('invalid JSON-RPC frame', err);
          return;
        }
        try {
          const response = await handler(frame);
          if (response !== undefined) {
            await sendResponse(response);
          }
        } catch (err) {
          log(`handler error for method '${frame.method}'`, err);
        }
      };

      const enqueue = (line: string): void => {
        dispatchChain = dispatchChain.then(() => handleOne(line));
      };

      const onData = (chunk: Buffer | string): void => {
        buffer += typeof chunk === 'string' ? chunk : decoder.write(chunk);
        let newlineIdx = buffer.indexOf('\n');
        while (newlineIdx !== -1) {
          const line = buffer.slice(0, newlineIdx);
          buffer = buffer.slice(newlineIdx + 1);
          enqueue(line);
          newlineIdx = buffer.indexOf('\n');
        }
      };

      const finish = async (): Promise<void> => {
        if (closed) return;
        closed = true;
        // Flush any bytes the decoder held while waiting for the rest of a
        // multi-byte sequence; emit them as part of the final line.
        buffer += decoder.end();
        if (buffer.length > 0) {
          enqueue(buffer);
          buffer = '';
        }
        // Drain serialized dispatch chain before resolving.
        await dispatchChain;
        resolve();
      };

      const onError = (err: Error): void => {
        if (closed) return;
        closed = true;
        reject(err);
      };

      input.on('data', onData);
      input.once('end', () => {
        void finish();
      });
      input.once('close', () => {
        void finish();
      });
      input.once('error', onError);
    });

  return { sendResponse, sendNotification, notify, run };
}

/** Build an `ok` response for the given request id. */
export function okResponse(id: RpcId | undefined, result: unknown): RpcResponse {
  return { jsonrpc: '2.0', id: id ?? null, result };
}

/** Build an `error` response for the given request id. */
export function errorResponse(id: RpcId | undefined, code: number, message: string, data?: unknown): RpcResponse {
  const error = data === undefined ? { code, message } : { code, message, data };
  return { jsonrpc: '2.0', id: id ?? null, error };
}

File: sdk/python/animus_plugin_sdk/animus_plugin_sdk/wire.py
"""Newline-delimited JSON-RPC 2.0 transport over stdio.

Mirrors the Rust runtime in `crates/animus-plugin-runtime/src/lib.rs`
and the TS SDK's `wire.ts`:

  - one JSON-RPC frame per line of stdin
  - replies + server-streaming notifications are newline-terminated JSON
    written to stdout
  - parse errors on a single line are logged-and-skipped (do not kill the
    read loop)
  - empty/whitespace-only lines are ignored

Stdout is reserved for protocol frames; SDK diagnostics MUST go to stderr.

Sync API (asyncio is overkill for stdio dispatch and complicates the SDK).
Pattern: `for line in sys.stdin: dispatch(parse(line))`.
"""

from __future__ import annotations

import json
import sys
from collections.abc import Callable, Iterable
from dataclasses import dataclass
from typing import IO, Any

from .types import ErrorCode, RpcError, RpcId, RpcNotification, RpcRequest, RpcResponse

FrameHandler = Callable[[RpcRequest], "RpcResponse | None"]
"""Handler invoked for every parsed inbound JSON-RPC frame.

Return `None` for notifications (no response expected).
"""

Logger = Callable[[str, BaseException | None], None]


def _default_logger(msg: str, err: BaseException | None = None) -> None:
    if err is not None:
        sys.stderr.write(f"[animus-plugin-sdk] {msg}: {err!s}\n")
    else:
        sys.stderr.write(f"[animus-plugin-sdk] {msg}\n")


def encode_frame(frame: RpcResponse | RpcNotification | RpcRequest) -> str:
    """Encode a single JSON-RPC frame as a newline-terminated string.

    Exposed for tests; production code should go through `Wire.send_response`
    or `Wire.send_notification`.

    JSON-RPC 2.0 requires:
      - response frames: exactly one of `result` or `error`; `id` always present
        (including `null` when the original request id was `null`).
      - notification frames: no `id`.
      - request frames: `id` present (including `null`) when expecting a reply.

    We therefore drop the unused half of a response (e.g. omit `error` on an
    `ok` reply) but always preserve `id` and the populated half, even when
    they are `null` — `exclude_none` alone would corrupt the wire shape.
    """
    payload: dict[str, Any] = {"jsonrpc": "2.0"}
    if isinstance(frame, RpcResponse):
        # Always emit `id` (possibly null). Emit exactly one of `result` /
        # `error` based on which was set; preserve `result: null` verbatim
        # (e.g. `subject/next` returning no subject).
        payload["id"] = frame.id
        if frame.error is not None:
            err_dict = frame.error.model_dump(exclude_none=True)
            payload["error"] = err_dict
        else:
            payload["result"] = frame.result
    elif isinstance(frame, RpcNotification):
        payload["method"] = frame.method
        if frame.params is not None:
            payload["params"] = frame.params
    else:  # RpcRequest
        payload["method"] = frame.method
        if frame.id is not None or "id" in frame.model_fields_set:
            payload["id"] = frame.id
        if frame.params is not None:
            payload["params"] = frame.params
    # Preserve any extra fields the wire types allow (`extra="allow"`).
    extras = frame.model_extra or {}
    for k, v in extras.items():
        if k not in payload:
            payload[k] = v
    return json.dumps(payload, separators=(",", ":")) + "\n"


def parse_frame(line: str) -> RpcRequest:
    """Parse one trimmed line into an `RpcRequest`.

    Raises `ValueError` on invalid JSON or structurally invalid JSON-RPC
    (missing `jsonrpc: "2.0"` or `method`).
    """
    try:
        value = json.loads(line)
    except json.JSONDecodeError as exc:
        raise ValueError(f"invalid JSON: {exc}") from exc
    if not isinstance(value, dict):
        raise ValueError("frame is not a JSON object")
    if value.get("jsonrpc") != "2.0":
        raise ValueError(f"unsupported jsonrpc version: {value.get('jsonrpc')!r}")
    method = value.get("method")
    if not isinstance(method, str) or not method:
        raise ValueError("frame missing string `method`")
    return RpcRequest.model_validate(value)


def ok_response(request_id: RpcId, result: Any) -> RpcResponse:
    """Build an `ok` response for the given request id."""
    return RpcResponse(jsonrpc="2.0", id=request_id, result=result, error=None)


def error_response(
    request_id: RpcId,
    code: int,
    message: str,
    data: Any | None = None,
) -> RpcResponse:
    """Build an `error` response for the given request id."""
    return RpcResponse(
        jsonrpc="2.0",
        id=request_id,
        result=None,
        error=RpcError(code=code, message=message, data=data),
    )


@dataclass
class Wire:
    """A bound stdio wire for sending frames + driving a read loop."""

    input: IO[str]
    output: IO[str]
    logger: Logger = _default_logger

    def send_response(self, response: RpcResponse) -> None:
        self.output.write(encode_frame(response))
        self.output.flush()

    def send_notification(self, notification: RpcNotification) -> None:
        self.output.write(encode_frame(notification))
        self.output.flush()

    def notify(self, method: str, params: Any | None = None) -> None:
        """Convenience: build + send a notification from method/params."""
        frame = RpcNotification(jsonrpc="2.0", method=method, params=params)
        self.send_notification(frame)

    def run(self, handler: FrameHandler) -> None:
        """Begin consuming the input stream until EOF.

        Frames that fail to parse are logged via `self.logger` and skipped,
        matching the Rust runtime's "tracing::warn + continue" behavior.
        """
        for line in self.input:
            self._dispatch_line(line, handler)

    # Internal helper extracted for unit testing.
    def _dispatch_line(self, line: str, handler: FrameHandler) -> None:
        trimmed = line.strip()
        if not trimmed:
            return
        try:
            frame = parse_frame(trimmed)
        except (ValueError, Exception) as exc:
            self.logger("invalid JSON-RPC frame", exc)
            return
        try:
            response = handler(frame)
        except Exception as exc:
            self.logger(f"handler error for method '{frame.method}'", exc)
            # Best-effort surface to host so request_id doesn't dangle.
            # Per JSON-RPC 2.0, ONLY a missing `id` makes a frame a
            # notification; `id: null` is still a request and the client is
            # waiting for a reply. Distinguish missing from explicit-null
            # via `model_fields_set`.
            if "id" in frame.model_fields_set:
                self.send_response(
                    error_response(
                        frame.id,
                        ErrorCode.INTERNAL_ERROR,
                        f"unhandled handler error: {exc!s}",
                    )
                )
            return
        if response is not None:
            self.send_response(response)


def create_wire(
    input: IO[str] | None = None,
    output: IO[str] | None = None,
    logger: Logger | None = None,
) -> Wire:
    """Create a `Wire` bound to the given (or default stdin/stdout) streams."""
    chosen_logger: Logger = logger if logger is not None else _default_logger
    return Wire(
        input=input if input is not None else sys.stdin,
        output=output if output is not None else sys.stdout,
        logger=chosen_logger,
    )


def run_lines(lines: Iterable[str], handler: FrameHandler, output: IO[str]) -> None:
    """Drive a sequence of pre-tokenized lines through the dispatcher.

    Useful in tests and for callers that have already line-buffered input.
    """
    wire = Wire(input=sys.stdin, output=output)
    for line in lines:
        wire._dispatch_line(line, handler)


__all__ = [
    "FrameHandler",
    "Logger",
    "Wire",
    "create_wire",
    "encode_frame",
    "error_response",
    "ok_response",
    "parse_frame",
    "run_lines",
]
